pkg/operators/operators.go

```diff
@@ -0,0 +1,333 @@
+	// Name must return a unique name for the operator
+	Name() string
+
+	// Description is an optional description to show to the user
+	Description() string
+
+	// GlobalParamDescs will return global params (required) for this operator
+	GlobalParamDescs() params.ParamDescs
+
+	// ParamDescs will return params (required) per gadget instance of the operator
+	ParamDescs() params.ParamDescs
+
+	// Dependencies can list other operators that this operator depends on
+	Dependencies() []string
+
+	// CanOperateOn should test whether the operator supports the given gadget. Init has not
+	// necessarily been called at this point.
+	CanOperateOn(gadgets.GadgetDesc) bool
+
+	// Init allows the operator to initialize itself
+	Init(params *params.Params) error
+
+	// Close allows the operator to clean up stuff prior to exiting
+	Close() error
+
+	// Instantiate is called before a gadget is run with this operator.
+	// This must return something that implements OperatorInstance.
+	// This is useful to create a context for an operator by wrapping it.
+	// Params given here are the ones returned by ParamDescs()
+	Instantiate(gadgetContext GadgetContext, gadgetInstance any, params *params.Params) (OperatorInstance, error)
+}
+
+type OperatorInstance interface {
+	// Name returns the name of the operator instance
+	Name() string
+
+	// PreGadgetRun in called before a gadget is run
+	PreGadgetRun() error
+
+	// PostGadgetRun is called after a gadget is run
+	PostGadgetRun() error
+
+	// Enricher should return a function that is able to operator on the event ev and
+	// call next afterwards.
+	Enricher(next EnricherFunc) EnricherFunc
+
+	EnrichEvent(ev any) error
+}
+
+type Operators []Operator
+
+// ContainerInfoFromMountNSID is a typical kubernetes operator interface that adds node, pod, namespace and container
+// information given the MountNSID
+type ContainerInfoFromMountNSID interface {
+	ContainerInfoSetters
+	GetMountNSID() uint64
+}
+
+type ContainerInfoSetters interface {
+	SetContainerInfo(pod, namespace, container string)
+	SetNode(string)
+}
+
+var allOperators = map[string]Operator{}
+
+type operatorWrapper struct {
+	Operator
+	initOnce    sync.Once
+	initialized bool
+}
+
+func (e *operatorWrapper) Init(params *params.Params) (err error) {
+	e.initOnce.Do(func() {
+		err = e.Operator.Init(params)
+		e.initialized = true
+	})
+	return err
+}
+
+// Register adds a new operator to the registry
+func Register(operator Operator) {
+	if _, ok := allOperators[operator.Name()]; ok {
+		panic(fmt.Errorf("operator already registered: %q", operator.Name()))
+	}
+	allOperators[operator.Name()] = operator
+}
+
+// GlobalParamsCollection returns a collection of params of all registered operators
+func GlobalParamsCollection() params.Collection {
+	pc := make(params.Collection)
+	for _, operator := range allOperators {
+		pc[operator.Name()] = operator.GlobalParamDescs().ToParams()
+	}
+	return pc
+}
+
+// GetAll returns a copy of all registered operators
```

Comment:
```suggestion
// GetAll returns all registered operators
```

I think copy can be misleading here as somebody could think that the copy means the object is deep-copied and the calls are in different instances of it, but IIUC the operators are always the same and not "real copies" are performed.